File: BEditor/BEditor/BasePropertyChanged.cpp
#include "pch.h"

using namespace System;
using namespace System::ComponentModel;
using namespace System::Diagnostics;
using namespace System::Linq;
using namespace System::Linq::Expressions;
using namespace System::Runtime::Serialization;
using namespace BEditor::Data;

generic<typename T> void BasePropertyChanged::SetValue(T src, T% dst, String^ name) {
	if (src == nullptr || !src->Equals(dst)) {
		dst = src;
		RaisePropertyChanged(name);
	}
}

void BasePropertyChanged::RaisePropertyChanged(String^ name) {
	this->PropertyChanged(this, gcnew PropertyChangedEventArgs(name));
}

File: BEditor/BEditor.NET/Data/BasePropertyChanged.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.Linq;
using System.Linq.Expressions;
using System.Runtime.Serialization;
using System.Text;

namespace BEditor.NET.Data {

    /// <summary>
    /// プロパティの変更を通知するクラス
    /// </summary>
    [DataContract(Namespace = "")]
    public abstract class BasePropertyChanged : INotifyPropertyChanged {

        public event PropertyChangedEventHandler PropertyChanged;

        protected void SetValue<T>(T src, ref T dst, string name) {
            if (src == null || !src.Equals(dst)) {
                dst = src;
                RaisePropertyChanged(name);
            }
        }

        protected void RaisePropertyChanged(string name) {
            if (PropertyChanged == null) return;

            PropertyChanged(this, new PropertyChangedEventArgs(name));
        }
    }
}
